File: src/ios/CordovaHotUpdate.m
//
//  CordovaHotUpdate.m
//  Cordova Hot Update Plugin
//

#import "CordovaHotUpdate.h"
#import "UpdateManager.h"

@interface CordovaHotUpdate ()

@property (nonatomic, strong) UpdateManager *updateManager;
@property (nonatomic, strong) NSUserDefaults *prefs;

@end

@implementation CordovaHotUpdate

- (void)pluginInitialize {
    [super pluginInitialize];
    self.prefs = [NSUserDefaults standardUserDefaults];
    self.updateManager = [[UpdateManager alloc] initWithPrefs:self.prefs];
}

- (void)initialize:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        @try {
            NSDictionary *options = [command.arguments objectAtIndex:0];
            NSString *appId = options[@"appId"];
            NSString *apiKey = options[@"apiKey"];
            NSString *baseUrl = options[@"baseUrl"] ?: @"https://cordova-hot-update.com";

            if (!appId || !apiKey) {
                CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                            messageAsString:@"appId and apiKey are required"];
                [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
                return;
            }

            // Save credentials
            [self.prefs setObject:appId forKey:@"appId"];
            [self.prefs setObject:apiKey forKey:@"apiKey"];
            [self.prefs setObject:baseUrl forKey:@"baseUrl"];
            [self.prefs setBool:YES forKey:@"initialized"];
            [self.prefs synchronize];

            [self.updateManager setCredentialsWithAppId:appId apiKey:apiKey baseUrl:baseUrl];

            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK
                                                        messageAsString:@"Initialized successfully"];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];

        } @catch (NSException *exception) {
            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                        messageAsString:exception.reason];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        }
    }];
}

- (void)checkForUpdate:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        if (![self isInitialized]) {
            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                        messageAsString:@"Plugin not initialized"];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
            return;
        }

        @try {
            NSDictionary *manifest = [self.updateManager checkForUpdate];

            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK
                                                     messageAsDictionary:manifest];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];

        } @catch (NSException *exception) {
            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                        messageAsString:exception.reason];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        }
    }];
}

- (void)applyUpdate:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        if (![self isInitialized]) {
            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                        messageAsString:@"Plugin not initialized"];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
            return;
        }

        @try {
            BOOL success = [self.updateManager downloadAndApplyUpdate];

            if (success) {
                CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK
                                                            messageAsString:@"Update applied successfully"];
                [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
            } else {
                CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                            messageAsString:@"Failed to apply update"];
                [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
            }

        } @catch (NSException *exception) {
            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                        messageAsString:exception.reason];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        }
    }];
}

- (void)getCurrentVersion:(CDVInvokedUrlCommand*)command {
    NSString *version = [self.prefs stringForKey:@"currentVersion"] ?: @"base";

    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK
                                                messageAsString:version];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)enableErrorReporting:(CDVInvokedUrlCommand*)command {
    BOOL enable = [[command.arguments objectAtIndex:0] boolValue];

    [self.prefs setBool:enable forKey:@"errorReportingEnabled"];
    [self.prefs synchronize];

    NSString *message = enable ? @"Error reporting enabled" : @"Error reporting disabled";
    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK
                                                messageAsString:message];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)reportError:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        if (![self isInitialized]) {
            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                        messageAsString:@"Plugin not initialized"];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
            return;
        }

        BOOL enabled = [self.prefs boolForKey:@"errorReportingEnabled"];
        if (!enabled) {
            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK
                                                        messageAsString:@"Error reporting is disabled"];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
            return;
        }

        @try {
            NSString *message = [command.arguments objectAtIndex:0];
            NSString *stack = [command.arguments objectAtIndex:1];

            [self.updateManager reportErrorWithMessage:message stack:stack];

            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK
                                                        messageAsString:@"Error reported"];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];

        } @catch (NSException *exception) {
            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                        messageAsString:exception.reason];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        }
    }];
}

- (void)getStatus:(CDVInvokedUrlCommand*)command {
    NSDictionary *status = @{
        @"initialized": @([self isInitialized]),
        @"currentVersion": [self.prefs stringForKey:@"currentVersion"] ?: @"base",
        @"errorReportingEnabled": @([self.prefs boolForKey:@"errorReportingEnabled"]),
        @"lastCheckTime": @([self.prefs doubleForKey:@"lastCheckTime"])
    };

    CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK
                                             messageAsDictionary:status];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)resetToBase:(CDVInvokedUrlCommand*)command {
    [self.commandDelegate runInBackground:^{
        @try {
            [self.updateManager resetToBase];

            [self.prefs setObject:@"base" forKey:@"currentVersion"];
            [self.prefs synchronize];

            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK
                                                        messageAsString:@"Reset to base version"];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];

        } @catch (NSException *exception) {
            CDVPluginResult *result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR
                                                        messageAsString:exception.reason];
            [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
        }
    }];
}

- (BOOL)isInitialized {
    return [self.prefs boolForKey:@"initialized"];
}

@end

File: src/android/CordovaHotUpdate.java
package com.cordova.hotupdate;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.content.Context;
import android.content.SharedPreferences;

/**
 * Cordova Hot Update Plugin - Main plugin class for Android
 */
public class CordovaHotUpdate extends CordovaPlugin {

    private static final String TAG = "CordovaHotUpdate";
    private static final String PREFS_NAME = "CordovaHotUpdatePrefs";

    private UpdateManager updateManager;
    private SharedPreferences prefs;

    @Override
    protected void pluginInitialize() {
        super.pluginInitialize();
        Context context = cordova.getActivity().getApplicationContext();
        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        updateManager = new UpdateManager(context, prefs);
    }

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {

        switch (action) {
            case "initialize":
                return this.initialize(args.getJSONObject(0), callbackContext);

            case "checkForUpdate":
                return this.checkForUpdate(callbackContext);

            case "applyUpdate":
                return this.applyUpdate(callbackContext);

            case "getCurrentVersion":
                return this.getCurrentVersion(callbackContext);

            case "enableErrorReporting":
                return this.enableErrorReporting(args.getBoolean(0), callbackContext);

            case "reportError":
                return this.reportError(args.getString(0), args.getString(1), callbackContext);

            case "getStatus":
                return this.getStatus(callbackContext);

            case "resetToBase":
                return this.resetToBase(callbackContext);

            case "loadStoredUpdate":
                return this.loadStoredUpdate(callbackContext);

            default:
                return false;
        }
    }

    /**
     * Initialize the plugin with app credentials
     */
    private boolean initialize(JSONObject options, CallbackContext callbackContext) {
        try {
            String appId = options.getString("appId");
            String apiKey = options.getString("apiKey");
            String baseUrl = options.optString("baseUrl", "https://cordova-hot-update.com");

            // Save credentials
            SharedPreferences.Editor editor = prefs.edit();
            editor.putString("appId", appId);
            editor.putString("apiKey", apiKey);
            editor.putString("baseUrl", baseUrl);
            editor.putBoolean("initialized", true);
            editor.apply();

            updateManager.setCredentials(appId, apiKey, baseUrl);

            callbackContext.success("Initialized successfully");
            return true;

        } catch (JSONException e) {
            callbackContext.error("Invalid options: " + e.getMessage());
            return false;
        }
    }

    /**
     * Check if a new update is available
     */
    private boolean checkForUpdate(CallbackContext callbackContext) {
        if (!isInitialized()) {
            callbackContext.error("Plugin not initialized");
            return true;
        }

        cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    JSONObject result = updateManager.checkForUpdate();
                    callbackContext.success(result);
                } catch (Exception e) {
                    callbackContext.error("Failed to check for update: " + e.getMessage());
                }
            }
        });

        return true;
    }

    /**
     * Download and apply the available update
     */
    private boolean applyUpdate(CallbackContext callbackContext) {
        if (!isInitialized()) {
            callbackContext.error("Plugin not initialized");
            return true;
        }

        cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    boolean success = updateManager.downloadAndApplyUpdate();
                    if (success) {
                        // Execute the hotfix immediately
                        String updateFilePath = prefs.getString("updateFilePath", null);
                        if (updateFilePath != null) {
                            java.io.File updateFile = new java.io.File(updateFilePath);
                            if (updateFile.exists()) {
                                // Read the file content
                                java.io.FileInputStream fis = new java.io.FileInputStream(updateFile);
                                byte[] data = new byte[(int) updateFile.length()];
                                fis.read(data);
                                fis.close();

                                final String jsCode = new String(data, "UTF-8");

                                android.util.Log.d(TAG, "[CordovaHotUpdate] Executing hotfix immediately...");

                                // Execute on UI thread
                                cordova.getActivity().runOnUiThread(new Runnable() {
                                    @Override
                                    public void run() {
                                        webView.loadUrl("javascript:" + jsCode);
                                        android.util.Log.d(TAG, "[CordovaHotUpdate] Hotfix executed successfully");
                                    }
                                });
                            }
                        }

                        callbackContext.success("Update applied successfully");
                    } else {
                        callbackContext.error("Failed to apply update");
                    }
                } catch (Exception e) {
                    callbackContext.error("Failed to apply update: " + e.getMessage());
                }
            }
        });

        return true;
    }

    /**
     * Get the currently installed update version
     */
    private boolean getCurrentVersion(CallbackContext callbackContext) {
        String version = prefs.getString("currentVersion", "base");
        callbackContext.success(version);
        return true;
    }

    /**
     * Enable or disable automatic error reporting
     */
    private boolean enableErrorReporting(boolean enable, CallbackContext callbackContext) {
        SharedPreferences.Editor editor = prefs.edit();
        editor.putBoolean("errorReportingEnabled", enable);
        editor.apply();

        callbackContext.success("Error reporting " + (enable ? "enabled" : "disabled"));
        return true;
    }

    /**
     * Manually report an error
     */
    private boolean reportError(String message, String stack, CallbackContext callbackContext) {
        if (!isInitialized()) {
            callbackContext.error("Plugin not initialized");
            return true;
        }

        boolean enabled = prefs.getBoolean("errorReportingEnabled", false);
        if (!enabled) {
            callbackContext.success("Error reporting is disabled");
            return true;
        }

        cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    updateManager.reportError(message, stack);
                    callbackContext.success("Error reported");
                } catch (Exception e) {
                    callbackContext.error("Failed to report error: " + e.getMessage());
                }
            }
        });

        return true;
    }

    /**
     * Get plugin info and status
     */
    private boolean getStatus(CallbackContext callbackContext) {
        try {
            JSONObject status = new JSONObject();
            status.put("initialized", isInitialized());
            status.put("currentVersion", prefs.getString("currentVersion", "base"));
            status.put("errorReportingEnabled", prefs.getBoolean("errorReportingEnabled", false));
            status.put("lastCheckTime", prefs.getLong("lastCheckTime", 0));

            callbackContext.success(status);
            return true;
        } catch (JSONException e) {
            callbackContext.error("Failed to get status: " + e.getMessage());
            return false;
        }
    }

    /**
     * Clear cached updates and reset to base version
     */
    private boolean resetToBase(CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    updateManager.resetToBase();

                    SharedPreferences.Editor editor = prefs.edit();
                    editor.putString("currentVersion", "base");
                    editor.apply();

                    callbackContext.success("Reset to base version");
                } catch (Exception e) {
                    callbackContext.error("Failed to reset: " + e.getMessage());
                }
            }
        });

        return true;
    }

    /**
     * Load and return the stored update from disk
     */
    private boolean loadStoredUpdate(CallbackContext callbackContext) {
        cordova.getThreadPool().execute(new Runnable() {
            @Override
            public void run() {
                try {
                    // Get the stored update file path
                    String updateFilePath = prefs.getString("updateFilePath", null);

                    if (updateFilePath == null) {
                        // No update stored
                        callbackContext.success("");
                        return;
                    }

                    // Check if file exists
                    java.io.File updateFile = new java.io.File(updateFilePath);
                    if (!updateFile.exists()) {
                        android.util.Log.d(TAG, "Update file not found at path: " + updateFilePath);
                        callbackContext.success("");
                        return;
                    }

                    // Read the file content
                    java.io.FileInputStream fis = new java.io.FileInputStream(updateFile);
                    byte[] data = new byte[(int) updateFile.length()];
                    fis.read(data);
                    fis.close();

                    String jsCode = new String(data, "UTF-8");

                    android.util.Log.d(TAG, "Loaded update from: " + updateFilePath);

                    // Return the JavaScript code
                    callbackContext.success(jsCode);

                } catch (Exception e) {
                    android.util.Log.e(TAG, "Exception loading update: " + e.getMessage());
                    callbackContext.error("Failed to load update: " + e.getMessage());
                }
            }
        });

        return true;
    }

    /**
     * Check if plugin is initialized
     */
    private boolean isInitialized() {
        return prefs.getBoolean("initialized", false);
    }
}
